dimod/generators/mimo.py

```diff
@@ -0,0 +1,1058 @@
+            parameter.
+
+        SNRb: Signal-to-noise ratio per bit, :math:`SNRb=10^{SNR_b[decibels]/10}`, 
+            used to generate the noisy signal when ``y`` is not provided. 
+            If ``float('Inf')``, no noise is added.  
+        
+        seed: Random seed, as an integer, or state, as a 
+            :class:`numpy.random.RandomState` instance.
+        
+        F_distribution: Zero-mean, variance-one distribution, in tuple form 
+            ``(distribution, type)``, used to generate each element in ``F`` 
+            when ``F`` is not provided . Supported values are:
+            
+            * ``'normal'`` or ``'binary'`` for the distribution 
+            * ``'real'`` or ``'complex'`` for the type 
+            
+            For large numbers of receivers and transmitters, statistical 
+            properties of the likelihood are weakly dependent on the 
+            distribution. Choosing ``'binary'`` allows for integer-valued 
+            Hamiltonians while ``'normal'`` is a more typical model. The channel 
+            can be real or complex; in many cases this represents a superficial 
+            distinction up to rescaling. For real-valued symbols (BPSK) the 
+            default is ``('normal', 'real')``; otherwise, the default is 
+            ``('normal', 'complex')``.
+        
+        use_offset: Adds a constant to the Ising model energy so that the 
+            energy evaluated for the transmitted symbols is zero. At 
+            sufficiently high ratios of receivers to users, and with high 
+            signal-to-noise ratio, this is with high probability the 
+            ground-state energy.
+
+    Returns:
+        bqm: Binary quadratic model defining the log-likelihood function.
+
+    Example:
+
+        Generate an instance of a CDMA problem in the high-load regime, near a
+        first-order phase transition:
+
+        .. doctest::                    # TODO: reconsider example/default graph 
+            :skipif: True
+
+            >>> import networkx as nx       
+            >>> G = nx.complete_graph(4)    
+            >>> nx.set_node_attributes(G, values={n:2*n+1 for n in G.nodes()}, name='num_transmitters')
+            >>> nx.set_node_attributes(G, values={n:2 for n in G.nodes()}, name='num_receivers')
+            >>> transmitted_symbols = np.random.choice([1, -1], 
+            ...     size=(sum(nx.get_node_attributes(G, "num_transmitters").values()), 1))
+            >>> bqm = dimod.generators.spin_encoded_comp(G,
+            ...     modulation='BPSK', 
+            ...     transmitted_symbols=transmitted_symbols,
+            ...     SNRb=5,
+            ...     F_distribution = ('binary', 'real'))
+
+    """
+
+    if not hasattr(lattice, 'edges') or not hasattr(lattice, 'nodes'): # not nx.Graph:
+        raise ValueError('Lattice must be a :class:`networkx.Graph`')
+
+    if modulation is None:
+        modulation = 'BPSK'
+
```

Comment:
Probably should just set the default `modulation` to `"BPSK"` so it's obvious that it's optional.